main/src/ca/uwaterloo/flix/language/phase/ControlFlowAnalysis.scala

```diff
@@ -0,0 +1,1787 @@
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.CompilationError
+import ca.uwaterloo.flix.language.ast.SimplifiedAst._
+import ca.uwaterloo.flix.language.ast.{SimplifiedAst, SourceLocation, Symbol, Type}
+import ca.uwaterloo.flix.language.debug.PrettyPrinter
+import ca.uwaterloo.flix.util.Validation._
+import ca.uwaterloo.flix.util.vt.TerminalContext
+import ca.uwaterloo.flix.util.{InternalCompilerException, Validation}
+
+object ControlFlowAnalysis extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+
+  type AbstractCache = Map[LabeledExpression, Set[Expression.Closure]]
+  type AbstractEnvironment = Map[Symbol.VarSym, Set[Expression.Closure]]
+
+  def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = flix.phase("CFA") {
+    var label = 0
```

Comment:
This is really not functional programming. The int should be passed as an argument.